### PR TITLE
Add CRootsOf incrementally for eigenvalues

### LINKED ISSUE #25282: incorrect multiplicities from `eigenvals`
To reproduce:
```
import numpy as np
import sympy as sp

print(sp.__version__)

dd = sd = [0] * 11 + [1]
ds = [2,0,1,0,0,0,1,0,1,0,1,0]
ss = ds.copy()
ss[8] = 2

def rotate(x, i):
    return x[i:] + x[:i]

mat = []
for i in range(12):
    mat.append(rotate(ss,i) + rotate(sd,i))
for i in range(12):
    mat.append(rotate(ds,i) + rotate(dd,i))

print(np.array(mat))

mat = sp.Matrix(mat)
eigenvalues = mat.eigenvals()
print('eigenvalues from eigenvals:')
for i in eigenvalues:
    print(eigenvalues[i], 'x', i)
print('matrix dimensions:', mat.shape)
print('sum of multiplicities:', sum([eigenvalues[i] for i in eigenvalues]))

eigenvectors = mat.eigenvects()
print('eigenvalues from eigenvects (multiplicities appear to be correct):')
for val, mult, _ in eigenvectors:
    print(mult, 'x', val)
```

Output:
```
1.12
[[2 0 1 0 0 0 1 0 2 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1]
 [0 1 0 0 0 1 0 2 0 1 0 2 0 0 0 0 0 0 0 0 0 0 1 0]
 [1 0 0 0 1 0 2 0 1 0 2 0 0 0 0 0 0 0 0 0 0 1 0 0]
 [0 0 0 1 0 2 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0]
 [0 0 1 0 2 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 1 0 2 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0]
 [1 0 2 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0]
 [0 2 0 1 0 2 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0]
 [2 0 1 0 2 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0]
 [0 1 0 2 0 1 0 0 0 1 0 2 0 0 1 0 0 0 0 0 0 0 0 0]
 [1 0 2 0 1 0 0 0 1 0 2 0 0 1 0 0 0 0 0 0 0 0 0 0]
 [0 2 0 1 0 0 0 1 0 2 0 1 1 0 0 0 0 0 0 0 0 0 0 0]
 [2 0 1 0 0 0 1 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1]
 [0 1 0 0 0 1 0 1 0 1 0 2 0 0 0 0 0 0 0 0 0 0 1 0]
 [1 0 0 0 1 0 1 0 1 0 2 0 0 0 0 0 0 0 0 0 0 1 0 0]
 [0 0 0 1 0 1 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0]
 [0 0 1 0 1 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 1 0 1 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0]
 [1 0 1 0 1 0 2 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0]
 [0 1 0 1 0 2 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0]
 [1 0 1 0 2 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0]
 [0 1 0 2 0 1 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0]
 [1 0 2 0 1 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0]
 [0 2 0 1 0 0 0 1 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 0]]
eigenvalues from eigenvals:
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 0)
1 x CRootOf(lambda**4 - 5*lambda**2 + 1, 0)
1 x -1
1 x CRootOf(lambda**4 - 5*lambda**2 + 1, 1)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 1)
1 x CRootOf(lambda**2 - 6*lambda - 1, 0)
1 x CRootOf(lambda**2 - 8*lambda + 1, 0)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 2)
1 x CRootOf(lambda**4 - 5*lambda**2 + 1, 2)
1 x 1
1 x CRootOf(lambda**4 - 5*lambda**2 + 1, 3)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 3)
1 x CRootOf(lambda**2 - 6*lambda - 1, 1)
1 x CRootOf(lambda**2 - 8*lambda + 1, 1)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 4)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 5)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 6)
1 x CRootOf(lambda**8 - 10*lambda**6 + 15*lambda**4 - 10*lambda**2 + 1, 7)
matrix dimensions: (24, 24)
sum of multiplicities: 18
eigenvalues from eigenvects (multiplicities appear to be correct):
2 x -1
2 x 1
2 x -sqrt(5/2 - sqrt(21)/2)
2 x sqrt(5/2 - sqrt(21)/2)
1 x 3 - sqrt(10)
1 x 3 + sqrt(10)
1 x 4 - sqrt(15)
1 x sqrt(15) + 4
2 x -sqrt(sqrt(21)/2 + 5/2)
2 x sqrt(sqrt(21)/2 + 5/2)
1 x -cos(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2) - I*sin(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2)
1 x -cos(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2) + I*sin(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2)
1 x cos(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2) - I*sin(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2)
1 x cos(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2) + I*sin(atan(sqrt(-33 + 20*sqrt(3))/(5 - 2*sqrt(3)))/2)
1 x -sqrt(sqrt(3) + 5/2 + sqrt(33 + 20*sqrt(3))/2)
1 x sqrt(sqrt(3) + 5/2 + sqrt(33 + 20*sqrt(3))/2)
1 x -sqrt(-sqrt(33 + 20*sqrt(3))/2 + sqrt(3) + 5/2)
1 x sqrt(-sqrt(33 + 20*sqrt(3))/2 + sqrt(3) + 5/2)
```
The multiplicities from `eigenvals` function only sum up to 18, while the matrix is of dimension 24. The output from `eigenvects` appears to be correct (and it's impressive that it can solve the 8th degree equation symbolically).

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #25282

#### Release Notes

- matrices
  - Fixed some corner cases when `Matrix.eigenvals` gives wrong result when `multiplicity=False`.

### CODE REVIEW

File: sympy/matrices/eigen.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def _eigenvals_dict(
             f = charpoly.as_expr()
             x = charpoly.gen
             try:
-                eigs = {CRootOf(f, x, idx): 1 for idx in range(degree)}
+                eigs = Counter(CRootOf(f, x, idx) for idx in range(degree))
```
Comment (oscarbenjamin) [inline sympy/matrices/eigen.py:261]: Maybe it could be `charpoly.all_roots(multiple=False)`

**Review by oscarbenjamin — APPROVED** [review]
Looks good to me

**Comment by oscarbenjamin** [comment]
Maybe use `charpoly.all_roots(multiple=False)`

**Comment by alreadydone** [comment]
Thanks for the PR! There are CI failures, which however don't seem to be related to this PR.

Have you considered using factorization of the characteristic polynomial to compute the eigenvalues with their algebraic multiplicities? That approach seems much faster according to my test on my example:

```
from sympy import Matrix, symbols
from time import time

dd = sd = [0] * 11 + [1]
ds = [2, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0]
ss = ds.copy()
ss[8] = 2

def rotate(x, i):
    return x[i:] + x[:i]

mat = []
for i in range(12):
    mat.append(rotate(ss, i) + rotate(sd, i))
for i in range(12):
    mat.append(rotate(ds, i) + rotate(dd, i))
mat = Matrix(mat)

t = time()
factors = mat.charpoly(symbols('lambda')).factor_list()
print('charpoly.factor_list takes:', time()-t)

t = time()
eigenvalues = mat.eigenvals()
print('eigenvals takes:', time()-t)
```

Output:
```
charpoly.factor_list takes: 0.08691191673278809
eigenvals takes: 0.9043083190917969
```

Once we get the irreducible factors, we can populate the eigenvalue dict with roots of the irreducible factors (`CRoots(p, i)` for `i = 0, ..., d-1` for a factor `p` of degree `d`, or `x0` for a factor `x - x0` of degree 1) and it seems that that would match the current `eigenvals` output.

**Comment by sylee957** [comment]
The 2.7 failing backlog prevents the merges though

**Pull request merged**